### PR TITLE
Added custom int and float TIFF tags

### LINKED ISSUE #660: Save custom TIFF tags in LZW-compressed TIFFs
If I understand correctly, the recent changes (#369) have not made it possible to write custom TIFF tags. I would like to do the following:

```python
info = TiffImagePlugin.ImageFileDirectory()
CustomTagId = 37000
info[CustomTagId] = 6
info.tagtype[CustomTagId] = 3 # 'short' according to TiffTags.TYPES

img.save(pathname, tiffinfo = info)
```

This corresponds to "Defining Application Tags" in http://www.remotesensing.org/libtiff/addingtags.html.

### PR DESCRIPTION
**Pull request by radarhere**
Resolves #660

### CODE REVIEW

File: src/encode.c [hunk 1]
```diff
@@ -804,7 +804,13 @@ PyImaging_LibTiffEncoderNew(PyObject* self, PyObject* args)
     PyObject *dir;
     PyObject *key, *value;
     Py_ssize_t pos = 0;
-    int status;
+    int key_int, status, core, i;
+    // This list also exists in TiffTags.py
+    const int tags[32] = {
```
Comment (hugovk) [inline src/encode.c]: Do we need to specify the size of the array?
```suggestion
    const int tags[] = {
```

File: src/encode.c [hunk 2]
```diff
@@ -845,29 +851,41 @@ PyImaging_LibTiffEncoderNew(PyObject* self, PyObject* args)
 
     for (pos = 0; pos < d_size; pos++) {
         key = PyList_GetItem(keys, pos);
+        key_int = (int)PyInt_AsLong(key);
         value = PyList_GetItem(values, pos);
         status = 0;
-        TRACE(("Attempting to set key: %d\n", (int)PyInt_AsLong(key)));
+        core = 0;
+        for (i=0; i<32; i++) {
```
Comment (hugovk) [inline src/encode.c]: Maybe remove the hardcoded array size?
```suggestion
        int number_of_tags = sizeof(tags) / sizeof(int);
        for (i=0; i<number_of_tags; i++) {
```

File: src/encode.c [hunk 3]
```diff
@@ -845,29 +851,41 @@ PyImaging_LibTiffEncoderNew(PyObject* self, PyObject* args)
 
     for (pos = 0; pos < d_size; pos++) {
         key = PyList_GetItem(keys, pos);
+        key_int = (int)PyInt_AsLong(key);
         value = PyList_GetItem(values, pos);
         status = 0;
-        TRACE(("Attempting to set key: %d\n", (int)PyInt_AsLong(key)));
+        core = 0;
```
Comment (hugovk) [inline src/encode.c]: Is this used to indicate that this tag is a core tag, and not a custom one?

Perhaps something like `is_core_tag` will be clearer, at first glance the `if (core || !ImagingLibTiffMergeFieldInfo(&encoder->state, TIFF_DOUBLE, key_int))` below looked like it's checking for PIL imaging core.

File: src/libImaging/TiffDecode.c [hunk 4]
```diff
@@ -336,6 +336,27 @@ int ImagingLibTiffEncodeInit(ImagingCodecState state, char *filename, int fp) {
...
+        { key, 0, 1, field_type, FIELD_CUSTOM, 1, 0, field_name }
+    };
+    n = sizeof(info) / sizeof(info[0]);
+
+    // Test for libtiff 4.0 or later, excluding libtiff 3.9.6 and 3.9.7
+    #if TIFFLIB_VERSION >= 20111221 && TIFFLIB_VERSION != 20120218 && TIFFLIB_VERSION != 20120922
+    status = TIFFMergeFieldInfo(clientstate->tiff, info, n);
+    #else
+    TIFFMergeFieldInfo(clientstate->tiff, info, n);
+    #endif
```
[the remark this answers is not in the document]
Comment (radarhere) [reply src/libImaging/TiffDecode.c]: Thanks, changed

File: src/libImaging/TiffDecode.c [hunk 5]
```diff
@@ -336,6 +336,27 @@ int ImagingLibTiffEncodeInit(ImagingCodecState state, char *filename, int fp) {
 
 }
 
+int ImagingLibTiffMergeFieldInfo(ImagingCodecState state, TIFFDataType field_type, int key){
+	TIFFSTATE *clientstate = (TIFFSTATE *)state->context;
+	char field_name[10];
+	uint32 n;
+	int status = 0;
+
+	sprintf(field_name, "Tag %d", key);
```
[the remark this answers is not in the document]
Comment (radarhere) [reply src/libImaging/TiffDecode.c]: Thanks for catching. I've now removing this debugging line.

File: src/libImaging/TiffDecode.c [hunk 6]
```diff
@@ -336,6 +336,27 @@ int ImagingLibTiffEncodeInit(ImagingCodecState state, char *filename, int fp) {
 
 }
 
+int ImagingLibTiffMergeFieldInfo(ImagingCodecState state, TIFFDataType field_type, int key){
```
[the remark this answers is not in the document]
Comment (radarhere) [reply src/libImaging/TiffDecode.c]: They were just in the lines that I added, but done.

File: src/encode.c (hunk 3 above)
Comment (radarhere) [reply src/encode.c]: Fair enough. Done.

File: Tests/test_file_libtiff.py [hunk 7]
```diff
@@ -231,6 +232,32 @@ def test_additional_metadata(self):
 
         TiffImagePlugin.WRITE_LIBTIFF = False
 
+    def test_custom_metadata(self):
+        custom = {
+            37000: 4,
+            37001: 4.2
```
[the remark this answers is not in the document]
Comment (radarhere) [reply Tests/test_file_libtiff.py]: 37000 was the tag number from the issue. Other than that, it is special in that it is above 32767 - [the TIFF specification](https://www.itu.int/itudoc/itu-t/com16/tiff-fx/docs/tiff6.pdf) states that 'An organization might wish to store information meaningful to only that organization
in a TIFF file. Tags numbered 32768 or higher, sometimes called private
tags, are reserved for that purpose.'

**Pull request merged**